### PR TITLE
Split out `arrow-data` into a separate crate

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of https://github.com/apache/arrow-rs/issues/2594

# Rationale for this change
 

ArrayData represents the canonical representation of an arrow array, and is the language of the C Data interface. It therefore makes sense to split into its own crate. In theory this could form the basis for interoperability between different Rust array abstractions.

# What changes are included in this PR?

Splits out ArrayData and related structures into a separate crate.

# Are there any user-facing changes?

We can no longer implement `From<ArrayRef> for ArrayData`. In practice this only appears to be being used by pyarrow code, and can be replaced with calls to `make_array` or `Array::data()` as appropriate.

### CODE REVIEW

File: arrow/src/array/array_decimal.rs [hunk 1]
```diff
@@ -284,7 +284,11 @@ impl<T: DecimalType> DecimalArray<T> {
 
         // safety: self.data is valid DataType::Decimal as checked above
         let new_data_type = Self::TYPE_CONSTRUCTOR(precision, scale);
-        Ok(self.data().clone().with_data_type(new_data_type).into())
+        let data = self.data().clone().into_builder().data_type(new_data_type);
```
Comment (tustvold) [inline arrow/src/array/array_decimal.rs:287]: This method was previously pub(crate), I opted to just remove it in favour of the into_builder plumbing

File: arrow-data/src/bitmap.rs [hunk 2]
```diff
@@ -56,6 +56,10 @@ impl Bitmap {
         unsafe { bit_util::get_bit_raw(self.bits.as_ptr(), i) }
     }
 
+    pub fn buffer(&self) -> &Buffer {
```
Comment (tustvold) [inline arrow-data/src/bitmap.rs:59]: I intend to deprecate buffer_ref in a follow up PR

File: arrow-data/src/transform/mod.rs [hunk 3]
```diff
@@ -0,0 +1,672 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (tustvold) [inline arrow-data/src/transform/mod.rs:1]: This is moved from array/transform/mod.rs but with the tests split out into arrow/tests/array_transform.rs

File: arrow/src/array/array.rs [hunk 4]
```diff
@@ -358,6 +358,90 @@ pub trait ArrayAccessor: Array {
     unsafe fn value_unchecked(&self, index: usize) -> Self::Item;
 }
 
+impl PartialEq for dyn Array {
```
Comment (tustvold) [inline arrow/src/array/array.rs:361]: These are moved from arrow/equal/mod.rs

File: arrow/src/array/equal/mod.rs [hunk 5]
```diff
@@ -1,1464 +0,0 @@
-// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (tustvold) [inline arrow/src/array/equal/mod.rs:1]: This is moved to arrow-data/equal/mod.rs with the array equality moved to arrow/array/array.rs and the tests moved to arrow/tests/array_equal.rs

File: arrow-data/src/data.rs [hunk 6]
```diff
@@ -1771,1124 +1746,30 @@ mod tests {
     }
 
     #[test]
-    #[should_panic(
```
Comment (tustvold) [inline arrow-data/src/data.rs:1774]: These tests are moved to arrow/tests/array_validation.rs

File: arrow-data/src/equal/boolean.rs [hunk 7]
```diff
@@ -88,30 +88,3 @@ pub(super) fn boolean_equal(
         })
     }
 }
-
-#[cfg(test)]
```
Comment (tustvold) [inline arrow-data/src/equal/boolean.rs:92]: These tests are moved to arrow/tests/array_equal.rs

File: arrow-data/src/equal/list.rs [hunk 8]
```diff
@@ -149,52 +148,3 @@ pub(super) fn list_equal<T: OffsetSizeTrait>(
         })
     }
 }
-
-#[cfg(test)]
```
Comment (tustvold) [inline arrow-data/src/equal/list.rs:153]: These tests are moved to arrow/tests/array_equal.rs

File: arrow-data/src/equal/mod.rs [hunk 9]
```diff
@@ -0,0 +1,171 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (tustvold) [inline arrow-data/src/equal/mod.rs:1]: This is moved from arrow/array/equal/mod.rs

File: arrow-data/src/transform/mod.rs (hunk 3 above)
Comment (alamb) [reply arrow-data/src/transform/mod.rs:1]: I think moving out the tests is a great idea -- I wonder if we need to do anything more to make them discoverable 🤔

File: arrow-data/src/transform/variable_size.rs [hunk 10]
```diff
@@ -15,31 +15,33 @@
...
+fn extend_offset_values<T: ArrowNativeType + AsPrimitive<usize>>(
     buffer: &mut MutableBuffer,
     offsets: &[T],
     values: &[u8],
     start: usize,
     len: usize,
 ) {
-    let start_values = offsets[start].to_usize().unwrap();
-    let end_values = offsets[start + len].to_usize().unwrap();
+    let start_values = offsets[start].as_();
```
Comment (alamb) [inline arrow-data/src/transform/variable_size.rs:36]: TIL: https://docs.rs/num/latest/num/cast/trait.AsPrimitive.html#tymethod.as_


`as_()` is something I have not seen before. Fascinating.

File: arrow/src/array/array_decimal.rs (hunk 1 above)
Comment (alamb) [reply arrow/src/array/array_decimal.rs:287]: cc @liukun4515 -- I think it is fine

**Review by alamb — APPROVED** [review]
I can't say I reviewed every single line in this PR super carefully, but I did read them all and looked at the changes, and verified the tests are still present. 

I also like the way that several of the larger tests are moved into `arrow/tests/...` integration style tests. This is similar to what arrow2 does 👍  Nice

Do we think it is appropriate to run any benchmarks?

LGTM I say we ship it

**Comment by tustvold** [comment]
This does not appear to have a discernible impact on the benchmarks - some fluctuate as is to be expected when you're talking nanoseconds, but no major change in either direction

**Pull request merged**